### PR TITLE
test(RHTAPWATCH-1066): Unit tests for getWorkspacesWithAccess

### PR DESCRIPTION
**Pull request by hmariset**
Add unit test for workspace-manager function
getWorkspacesWithAccess

Jira-Url: https://issues.redhat.com/browse/RHTAPWATCH-1066

### CODE REVIEW

**Comment by hmariset** [comment]
@Omeramsc, @avi-biton  Please review the PR.

File: cmd/main_test.go [hunk 1]
```diff
@@ -301,3 +307,105 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+			namespaceNames := []string{"test-tenant", "test-tenant-2"}
+			for _, name := range namespaceNames {
+				ns, err := createNamespace(k8sClient, name)
+				Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Error while creating the namespace %s: %v", name, err))
+				allNamespaces = append(allNamespaces, ns)
+			}			
+			createRole(k8sClient, "test-tenant", "namespace-access", []string{"create", "list", "watch", "delete"})
+			createRole(k8sClient, "test-tenant-2", "namespace-access-2", []string{"create", "list", "watch", "delete"})
+			createRoleBinding(k8sClient, "namespace-access-user-binding", "test-tenant", "user1@konflux.dev", "namespace-access")
+			createRoleBinding(k8sClient, "namespace-access-user-binding-2", "test-tenant", "user2@konflux.dev", "namespace-access")
```
Comment (Omeramsc) [inline cmd/main_test.go]: You can add those assignments to the loop

File: cmd/main_test.go [hunk 2]
```diff
@@ -301,3 +307,105 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+			gv = "v1alpha1"
+			namespaceNames := []string{"test-tenant", "test-tenant-2"}
+			for _, name := range namespaceNames {
+				ns, err := createNamespace(k8sClient, name)
+				Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Error while creating the namespace %s: %v", name, err))
+				allNamespaces = append(allNamespaces, ns)
+			}			
+			createRole(k8sClient, "test-tenant", "namespace-access", []string{"create", "list", "watch", "delete"})
+			createRole(k8sClient, "test-tenant-2", "namespace-access-2", []string{"create", "list", "watch", "delete"})
+			createRoleBinding(k8sClient, "namespace-access-user-binding", "test-tenant", "user1@konflux.dev", "namespace-access")
```
Comment (Omeramsc) [inline cmd/main_test.go]: Is it relevant for this test to have both user1 and user2?

File: cmd/main_test.go [hunk 3]
```diff
@@ -301,3 +307,105 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+								Name: "test-tenant-2",
+								Type: "default",
+							},
+						},
+					},
+				},
+			}
+		})
+
+		It("WorkspaceList should be created with ns1 and ns2 names", func() {
```
Comment (Omeramsc) [inline cmd/main_test.go]: Let's try rephrasing it a tiny bit to make the schema more readable
something similar to this:
  Describe("GetWorkspacesWithAccess querying for workspaces with access"  ) {
Context("When workspace "test-tenant"'s namespaces has  all the necessary permissions",) { 
It("Should return a WorkspaceList with test-tenant workspace and ns1 and ns2 in it")

Comment (Omeramsc) [reply cmd/main_test.go]: (same for the rest of the tests)

File: cmd/main_test.go [hunk 4]
```diff
@@ -301,3 +307,105 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+			}
+		})
+
+		It("WorkspaceList should be created with only ns1 name", func() {
+			actualWorkspaces, err := getWorkspacesWithAccess(e, c, allNamespaces)
+			Expect(actualWorkspaces.Items).To(Equal(expectedWorkspaces))
+			Expect(err).NotTo(HaveOccurred(), "Unexpected error testing GetWorkspacesWithAccess")
+		})
+	})	
+})
```
Comment (Omeramsc) [inline cmd/main_test.go:478]: Add a case of no workspaces returned

**Review by Omeramsc — CHANGES_REQUESTED** [review]

File: cmd/main_test.go [hunk 5]
```diff
@@ -301,3 +307,122 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+	Context("When workspace test-tenant's namespaces has all the necessary permissions", func() {
+		BeforeEach(func() {
+			gv = "v1alpha1"
+			namespaceNames := []string{"test-tenant", "test-tenant-2"}
+			for _, name := range namespaceNames {
+				ns, err := createNamespace(k8sClient, name)
+				Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Error while creating the namespace %s: %v", name, err))
+				allNamespaces = append(allNamespaces, ns)
+				createRole(k8sClient, name, "namespace-access", []string{"create", "list", "watch", "delete"})
+				createRoleBinding(k8sClient, "namespace-access-user-binding", name, "user1@konflux.dev", "namespace-access")
```
Comment (Omeramsc) [inline cmd/main_test.go]: Those two lines should result in an error, as a resource with this name ( "namespace-access-user-binding", "namespace-access") already exists (as you are creating them for two namespaces). if it did not error, check out why. 

after you see why, you can use an index to create it in the loop.

File: cmd/main_test.go [hunk 6]
```diff
@@ -301,3 +307,122 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+			actualWorkspaces, err := getWorkspacesWithAccess(e, c, allNamespaces)
+			Expect(actualWorkspaces.Items).To(Equal(expectedWorkspaces))
+			Expect(err).NotTo(HaveOccurred(), "Unexpected error testing GetWorkspacesWithAccess")
+		})
+	})
+
+	Context("When workspace with only test-tenant namespace has all the necessary permissions", func() {
+		BeforeEach(func() {
+			gv = "v1alpha1"
+			namespaceNames := []string{"test-tenant", "test-tenant-3"}
```
Comment (Omeramsc) [inline cmd/main_test.go]: why are we using test-tenant-3 and not 2?

File: cmd/main_test.go [hunk 7]
```diff
@@ -301,3 +307,122 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+			for _, name := range namespaceNames {
+				ns, err := createNamespace(k8sClient, name)
+				Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Error while creating the namespace %s: %v", name, err))
+				allNamespaces = append(allNamespaces, ns)
+			}			
+			createRole(k8sClient, "test-tenant", "namespace-access", []string{"create", "list", "watch", "delete"})
```
Comment (Omeramsc) [inline cmd/main_test.go]: as we did not clean up the previous test, this should fail as we create the same resource twice.

File: cmd/main_test.go [hunk 8]
```diff
@@ -301,3 +307,122 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+			createRoleBinding(k8sClient, "namespace-access-user-binding", "test-tenant", "user1@konflux.dev", "namespace-access")
```
Comment (Omeramsc) [inline cmd/main_test.go]: As we did not clean up the previous test, how can we know this does not clash with the roles we binded to this namespace in the previous test?

File: cmd/main_test.go [hunk 9]
```diff
@@ -301,3 +307,122 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+			createRoleBinding(k8sClient, "namespace-access-user-binding-3", "test-tenant-3", "user1@konflux.dev", "namespace-access")
```
Comment (Omeramsc) [inline cmd/main_test.go]: I don't get it, we gave test-tenant and test-tenant-3 the same roles and permissions, why does the test return only test-tenant in the workspace?

File: cmd/main_test.go [hunk 10]
```diff
@@ -301,3 +307,122 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+	Context("When no workspaces has all the necessary permissions", func() {
+		BeforeEach(func() {
+			gv = "v1alpha1"
+			namespaceNames := []string{"test-tenant", "test-tenant-2"}
+			for _, name := range namespaceNames {
+				ns, err := createNamespace(k8sClient, name)
+				Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Error while creating the namespace %s: %v", name, err))
+				allNamespaces = append(allNamespaces, ns)
+			}			
+			expectedWorkspaces = []crt.Workspace{}
```
Comment (Omeramsc) [inline cmd/main_test.go:461]: I'm not sure how that is the result we are getting, taking into considiration we did not clean up the resources we created on the other test cases

**Review by Omeramsc — CHANGES_REQUESTED** [review]
Good progress, please add cleanup after the unit tests.

File: cmd/main_test.go [hunk 11]
```diff
@@ -301,3 +307,138 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+		})
+
+		It("Should return a empty WorkspaceList", func() {
+			actualWorkspaces, err := getWorkspacesWithAccess(e, c, allNamespaces)
+			Expect(actualWorkspaces.Items).To(Equal(expectedWorkspaces))
+			Expect(err).NotTo(HaveOccurred(), "Unexpected error testing GetWorkspacesWithAccess")
+		})
+
+		AfterEach(func() {
+			allNamespaces = []k8sapi.Namespace{}
```
Comment (Omeramsc) [inline cmd/main_test.go]: Those are not the variables you need to clean, 
what we need to clean are the actual resources.

for example:
in the test you ran the following- 
```
createNamespace(k8sClient, name)
createRole(k8sClient, name, "namespace-access", []string{"create", "list", "watch", "delete"})
createRoleBinding(k8sClient, "namespace-access-user-binding", name, "user1@konflux.dev", "namespace-access")
```

those resources may stay available on the envtest cluster while the other tests are running, as we use the same env. we should remove all resources we created.

**Review by Omeramsc — CHANGES_REQUESTED** [review]
resource managment between tests require fixing.

Also if my comments are not about typos/basic syntax, please comment on them with how you fixed them before resolving them (or wait for my resolution).

**Review by Omeramsc — APPROVED** [review]
LGTM

**Pull request merged**